File: js/step4_if_fn_do.js
var readline = require('./readline'),
    types = require('./types'),
    reader = require('./reader'),
    printer = require('./printer'),
    Env = require('./env').Env,
    core = require('./core')

// read
function READ(str) {
    return reader.read_str(str)
}

// eval
function eval_ast(ast, env) {
    if (ast instanceof types.Symbol) {
        return env.get(ast)
    } else if (Array.isArray(ast)) {
        return ast.map(function(a) { return EVAL(a, env) })
    } else {
        return ast
    }
}

function EVAL(ast, env) {
    if (!Array.isArray(ast)) {
        return eval_ast(ast, env)
    }
    if (ast.length === 0) { return ast }

    // apply list
    var a0 = ast[0], a1 = ast[1], a2 = ast[2], a3 = ast[3]
    switch (a0.name) {
    case 'def!':
        return env.set(a1, EVAL(a2, env))
    case 'let*':
        var let_env = new Env(env)
        for (var i=0; i < a1.length; i+=2) {
            let_env.set(a1[i], EVAL(a1[i+1], let_env))
        }
        return EVAL(a2, let_env)
    case "do":
        var el = eval_ast(ast.slice(1), env)
        return el[el.length-1]
    case "if":
        var cond = EVAL(a1, env)
        if (cond === null || cond === false) {
            return ast.length > 3 ? EVAL(a3, env) : null
        } else {
            return EVAL(a2, env)
        }
    case "fn*":
        return function() {
            return EVAL(a2, new Env(env, a1, arguments))
        }
    default:
        var el = eval_ast(ast, env), f = el[0]
        return f.apply(f, el.slice(1))
    }
}

// print
function PRINT(exp) {
    return printer.pr_str(exp, true)
}

// repl
repl_env = new Env(null)

// core.js: defined using javascript
for (var k in core.ns) { repl_env.set(new types.Symbol(k), core.ns[k]) }

// core.mal: defined using mal itself
rep("(def! not (fn* (a) (if a false true)))")

function rep (line) {
    return PRINT(EVAL(READ(line), repl_env))
}

while ((line = readline.readlineSync('user> ')) !== null) {
    try {
        process.stdout.write(rep(line) + '\n')
    } catch (exc) {
        console.log(exc.stack)
    }
}

File: python/step4_if_fn_do.py
import sys, traceback
import readline
import mal_types as types
import reader, printer
from env import Env
import core

# read
def READ(str):
    return reader.read_str(str)

# eval
def eval_ast(ast, env):
    if type(ast) == types.Symbol:
        return env.get(ast)
    elif type(ast) == list:
        return list(map(lambda a: EVAL(a, env), ast))
    else:
        return ast

def EVAL(ast, env):
    if not type(ast) == list:
        return eval_ast(ast, env)

    # apply list
    if len(ast) == 0: return ast
    a0 = ast[0]
    if "def!" == a0:
        return env.set(ast[1], EVAL(ast[2], env))
    elif "let*" == a0:
        let_env = Env(env)
        for i in range(0, len(ast[1]), 2):
            let_env.set(ast[1][i], EVAL(ast[1][i+1], let_env))
        return EVAL(ast[2], let_env)
    elif "do" == a0:
        el = eval_ast(ast[1:], env)
        return el[-1]
    elif "if" == a0:
        a1, a2 = ast[1], ast[2]
        cond = EVAL(a1, env)
        if cond is None or cond is False:
            if len(ast) > 3: return EVAL(ast[3], env)
            else:            return None
        else:
            return EVAL(a2, env)
    elif "fn*" == a0:
        return types.Function(ast[2], ast[1], env)
    else:
        el = eval_ast(ast, env)
        f = el[0]
        if type(f) == types.Function:
            return EVAL(f.ast, Env(f.env, f.params, el[1:]))
        else:
            return f(*el[1:])


# print
def PRINT(exp):
    return printer.pr_str(exp)

# repl
repl_env = Env()
def REP(str):
    return PRINT(EVAL(READ(str), repl_env))

# core.py: defined using python
for k,v in core.ns.items(): repl_env.set(types.Symbol(k), v)

# core.mal: defined using the language itself
REP("(def! not (fn* (a) (if a false true)))")

# repl loop
while True:
    try:
        line = input("user> ")
        if line == None: break
        if line == "": continue
        print(REP(line))
    except reader.Blank: continue
    except EOFError: break
    except Exception as e:
        print("".join(traceback.format_exception(*sys.exc_info())))
